### DISCUSSION TITLE
Audio Player dispose

### DISCUSSION THREAD

**Discussion by ardera** [bluefireteam/audioplayers#1746] [Q&A]
All audio players need to be disposed, right?

IIUC, `ReleaseMode.release` only releases the internal audio playback pipeline on EOS and only on android. So if `AudioPlayer.dispose` is not called, on linux, that'll leave the native audio player alive forever including the pipeline, correct? And on android, only the audio player, without the playback pipeline is kept alive.

So https://github.com/flame-engine/flame/blob/1d9fe6139287b984a05e0056c279b9c5d277e026/packages/flame_audio/example/lib/main.dart#L47-L49 should quickly lead to resource exhaustion. Is this correct?

**Comment by Gustl22** [comment]
Yes this might be correct. Therefore we should implement https://github.com/bluefireteam/audioplayers/issues/1689 in order to fix this.

**Comment by iatndiag** [comment]
In Linux if you will use Dispose() in Music App, you will give no sound at all. With every new Note you will give +1 Volume Regulator in Ubuntu -> Settings -> Sound window. Their number will grow endlessly and will cause sound stop.
The Solution for Linux musical App is to use Release() method. There will be 4...5 volume controls in the settings window. The fifth one will appear and disappear. This is the normal mode. Five is enough for the previous notes to sound to the end. Something like this:

    if(Platform.isLinux) {              // Prevention of endless Multiple instanses of Volume Regulators in Linux Sound Settings
        //if(audioPlayersMapLinux.containsKey(ntTblNtfrsList[24]['counterPlayerLinux'].toString())) {} else {                // do nothing
          tempCLinux = ntTblNtfrsList[24]['counterPlayerLinux'].toString();
          if(audioPlayersMapLinux.length < 5) {
            audioPlayersMapLinux.addEntries({tempCLinux : AudioPlayer()}.entries);  // adding AudioPlayer for absolutely new note, .stop() method !!!
          } else {
                      if(ntTblNtfrsList[24]['counterPlayerLinux'] == 4 && audioPlayersMapLinux.length == 5) {       // LINUX SOLUTION IS a RELEASE()    NOT Dispose()   !!!
                        tempCLinux_toRelease = '0';
              } else if (ntTblNtfrsList[24]['counterPlayerLinux'] == 3 && audioPlayersMapLinux.length == 5) {
                        tempCLinux_toRelease = '4';
              } else if (ntTblNtfrsList[24]['counterPlayerLinux'] == 2 && audioPlayersMapLinux.length == 5) {
                        tempCLinux_toRelease = '3';
              } else if (ntTblNtfrsList[24]['counterPlayerLinux'] == 1 && audioPlayersMapLinux.length == 5) {
                        tempCLinux_toRelease = '2';
              } else if (ntTblNtfrsList[24]['counterPlayerLinux'] == 0 && audioPlayersMapLinux.length == 5) {
                        tempCLinux_toRelease = '1';
              } else {}
          } //end if
          audioPlayersMapLinux[tempCLinux]?.setVolume(nVol);
        //await audioPlayersMapLinux[tempCLinux]?.stop(); await audioPlayersMapLinux[tempCLinux]?.play(AssetSource('$nT_'), mode: PlayerMode.lowLatency);
          await audioPlayersMapLinux[tempCLinux]?.stop(); await audioPlayersMapLinux[tempCLinux]?.play(AssetSource('$nT_'), mode: PlayerMode.lowLatency);
          if(audioPlayersMapLinux.length >= 5)  {
          await audioPlayersMapLinux[tempCLinux_toRelease]?.release();    // LINUX SOLUTION IS a RELEASE()    NOT Dispose()   !!!   Number of Volume Regulators Stops to grow!
        //await audioPlayersMapLinux[tempCLinux_toRelease]?.dispose();    // sound disappers after number of volume regulators overflow
          } else {}       // Linux Solution is DISPOSE()
          if(ntTblNtfrsList[24]['counterPlayerLinux']! < 4) {ntTblNtfrsList[24]['counterPlayerLinux'] = ntTblNtfrsList[24]['counterPlayerLinux']! + 1;} else {ntTblNtfrsList[24]['counterPlayerLinux'] = 0;}
          ntTblNotifier.value = ntTblNtfrsList;
          //await player.dispose();     // Linux added  DISPOSE() to delete volume Regulators (Sliders). Sounds stops Linux when Number of volume Regulators Overfilled !!!
          //player.release();
        } else {      // build mode: android
          await player.play(AssetSource('$nT_'), mode: PlayerMode.lowLatency);  // Android Sounds fine From Here, Mode for each note: lowLatency ! But Memory Leak Windows! Multiple Endless list of Volume Regulators Linux!
        }
